samples/nrf9160/modem_shell/src/at/at_cmd_mode.c

```diff
@@ -0,0 +1,344 @@
+};
+static enum term_modes term_mode;
+
+#define AT_CMD_MODE_AT_CMD_MAX_LEN CONFIG_SHELL_CMD_BUFF_SIZE
+#define AT_CMD_MODE_PIPELINED_AT_CMD_MAX_LEN 1023
+
+static uint8_t at_buf[AT_CMD_MODE_AT_CMD_MAX_LEN + 1];
+static uint8_t at_buf_pipelined[AT_CMD_MODE_PIPELINED_AT_CMD_MAX_LEN + 1];
+
+AT_MONITOR(mosh_at_cmd_mode_handler, ANY, at_cmd_mode_event_handler, PAUSED);
+
+
+struct at_cmd_mode_pipelining_info {
+	uint8_t *pipelined_cmd_start; /* Pointers to at_buf with pipelining */
+	int pipelined_cmd_len;
+};
+
+#define AT_MAX_CMD_PIPELINED 10
+struct at_cmd_mode_work_data {
+	struct k_work work;
+	int pipe_cnt;
+	struct at_cmd_mode_pipelining_info pipe_infos[AT_MAX_CMD_PIPELINED];
+};
+static struct at_cmd_mode_work_data at_cmd_mode_work;
+
+static bool at_cmd_mode_active;
+static bool inside_quotes;
+static bool writing;
+static size_t at_cmd_len;
+
+K_MUTEX_DEFINE(at_buf_mutex);
+
+static void at_cmd_mode_event_handler(const char *response)
+{
+	if (writing) {
+		/* 1st we need to clear current writings */
+		printk("\r");
+		for (int i = 0; i < at_cmd_len; i++) {
+			printk(" ");
+		}
+
+		/* Then print AT notification */
+		printk("\r%s", response);
+
+		/* And at last, print the command that user was writing back */
+		k_mutex_lock(&at_buf_mutex, K_FOREVER);
+		at_buf[at_cmd_len] = '\0';
+		printk("%s", at_buf);
+		k_mutex_unlock(&at_buf_mutex);
+	} else {
+		printk("%s", response);
+	}
+}
+
+static void at_cmd_mode_send_at_cmd(const char *at_str, char *rsp_buf, int rsp_buf_len)
+{
+	int err;
+
+	/* 1st: handle MoSh specific AT commands */
+	err = at_cmd_mosh_parse(at_str);
+	if (err == 0) {
+		sprintf(rsp_buf, "OK\n");
+		goto done;
+	} else if (err != -ENOENT) {
+		sprintf(rsp_buf, "ERROR\n");
+		goto done;
+	}
+
+	/* 2nd: modem at commands */
+	err = nrf_modem_at_cmd(rsp_buf, rsp_buf_len, "%s", at_str);
+	if (err < 0) {
+		sprintf(rsp_buf, "ERROR: AT command failed: %d\n", err);
+	} /* else if (err > 0): print error from modem */
+done:
+	/* Response */
+	printk("%s", rsp_buf);
+}
+
+static void at_cmd_mode_send_at_cmd_from_at_buf(void)
+{
+	at_cmd_mode_send_at_cmd(at_buf, at_buf, sizeof(at_buf));
+}
+
+static void at_cmd_mode_worker(struct k_work *item)
+{
+	int i;
+	struct at_cmd_mode_work_data *data = CONTAINER_OF(item, struct at_cmd_mode_work_data, work);
+	struct at_cmd_mode_pipelining_info *pipes = data->pipe_infos;
+
+	if (data->pipe_cnt) {
+		for (i = 0; i < AT_MAX_CMD_PIPELINED; i++) {
+			if (pipes[i].pipelined_cmd_len) {
+				strncpy(at_buf_pipelined, pipes[i].pipelined_cmd_start,
+					pipes[i].pipelined_cmd_len);
+				at_buf_pipelined[pipes[i].pipelined_cmd_len] = '\0';
+
+				at_cmd_mode_send_at_cmd(at_buf_pipelined, at_buf_pipelined,
+							sizeof(at_buf_pipelined));
+				pipes[i].pipelined_cmd_len = 0;
+			}
+		}
+	} else {
+		at_cmd_mode_send_at_cmd_from_at_buf();
+	}
+	data->pipe_cnt = 0;
+}
+
+static void at_cmd_mode_cmd_rx_handler(uint8_t character)
+{
+	writing = true;
+
+	/* Handle control characters */
```

Comment:
at_cmd_mode_cmd_rx_handler() is very long function. Consider placing for example this part to a separate function.